src/widgets/RealtimeEffectPanel.cpp

```diff
@@ -0,0 +1,814 @@
+            dropHintLine->SetSize(item->GetSize().x, 3);
+         }
+      });
+      Bind(EVT_MOVABLE_CONTROL_DRAG_FINISHED, [this, dropHintLine](const MovableControlEvent& event)
+      {
+         dropHintLine->Hide();
+         
+         auto trackOwner = mTrack->GetOwner();
+         if(!trackOwner)
+            return;
+
+         auto project = trackOwner->GetOwner();
+         if(project == nullptr)
+            return;
+
+         auto& effectList = RealtimeEffectList::Get(*mTrack);
+         auto from = event.GetSourceIndex();
+         auto to = event.GetSourceIndex() < event.GetTargetIndex() ? event.GetTargetIndex() - 1 : event.GetTargetIndex();
+         if(from != to)
+         {
+            effectList.MoveEffect(from, to);
+            ProjectHistory::Get(*project).PushState(
+               //i18n-hint: undo history, first parameter - track name
+               XO("'%s' effects reorder %d->%d").Format(mTrack->GetName(), from, to),
+               //i18n-hint: undo history record
+               XO("Realtime effect reorder"), UndoPush::CONSOLIDATE);
+         }
+         else
+         {
+            wxWindowUpdateLocker freeze(this);
+            Layout();
+         }
+      });
+      SetScrollRate(0, 20);
+   }
+
+   void OnEffectListItemChange(const RealtimeEffectListMessage& msg)
+   {
+      wxWindowUpdateLocker freeze(this);
+      if(msg.type == RealtimeEffectListMessage::Type::Move)
+      {
+         const auto sizer = mEffectListContainer->GetSizer();
+
+         const auto movedItem = sizer->GetItem(msg.index);
+
+         const auto proportion = movedItem->GetProportion();
+         const auto flag = movedItem->GetFlag();
+         const auto border = movedItem->GetBorder();
+
+         auto window = movedItem->GetWindow();
+         
+         sizer->Remove(msg.index);
+         sizer->Insert(std::any_cast<size_t>(msg.extra), window, proportion, flag, border);
+      }
+      else if(msg.type == RealtimeEffectListMessage::Type::Insert)
+      {
+         auto& effects = RealtimeEffectList::Get(*mTrack);
+         InsertEffectRow(msg.index, effects.GetStateAt(msg.index));
+      }
+      else if(msg.type == RealtimeEffectListMessage::Type::Remove)
+      {
+         auto sizer = mEffectListContainer->GetSizer();
+         auto item = sizer->GetItem(msg.index)->GetWindow();
+         item->Destroy();
+      }
+      Layout();
+   }
+
+   void ResetTrack()
+   {
+      mEffectListItemMovedSubscription.Reset();
+      mUndoManagerSubscription.Reset();
+
+      mTrack.reset();
+      ReloadEffectsList();
+   }
+
+   void SetTrack(AudacityProject& project, const std::shared_ptr<Track>& track)
+   {
+      if(mTrack == track)
+         return;
+
+      mEffectListItemMovedSubscription.Reset();
+      mUndoManagerSubscription.Reset();
+      
+      mTrack = track;
+      ReloadEffectsList();
+
+      if(track)
+      {
+         auto& effects = RealtimeEffectList::Get(*mTrack);
```

Comment:
Here and in `ReloadEffectList`, please check that the track is a WaveTrack before getting the `RealtimeEffectList`, just to avoid creation-on-demand of a useless, empty list in case the track is of non-wave type.

Reply:
Shouldn't be the case as if we do not show "effects" button for other types of tracks

Reply:
But changing track focus using the up and down arrow keys is another way to change the contents of the sidebar.

Reply:
Didn't think of that, thanks